### PR TITLE
refactor: example usage of ngx-speculoos

### PR DESCRIPTION
**Pull request by jnizet**
Not to be merged (yet).

This PR allows showing the usage on several real components of ngx-speculoos (https://github.com/Ninja-Squad/ngx-speculoos).

### CODE REVIEW

File: frontend/src/app/person-edit/person-edit.component.spec.ts [hunk 1]
```diff
@@ -133,10 +278,10 @@ describe('PersonEditComponent', () => {
     })));
 
     it('should have a title', () => {
-      const fixture = TestBed.createComponent(PersonEditComponent);
-      fixture.detectChanges();
+      const tester = new PersonEditTester(TestBed.createComponent(PersonEditComponent));
```
Comment (cexbrayat) [inline frontend/src/app/person-edit/person-edit.component.spec.ts]: Could that be simplified to: `new PersonEditTester(PersonEditComponent)`? Is there a reason to create and pass the fixture?

File: frontend/src/app/person-edit/person-edit.component.spec.ts [hunk 2]
```diff
@@ -145,76 +290,44 @@ describe('PersonEditComponent', () => {
...
-      const nationality = nativeElement.querySelector('#nationality');
-      expect(nationality.value).toBe(person.nationality.name);
-
-      lastName.value = 'Do';
-      lastName.dispatchEvent(new Event('input'));
-      nativeElement.querySelector('#save').click();
+      const tester = new PersonEditTester(TestBed.createComponent(PersonEditComponent));
+      tester.detectChanges();
+
+      expect(tester.firstName.value).toBe(person.firstName);
```
Comment (cexbrayat) [inline frontend/src/app/person-edit/person-edit.component.spec.ts]: Maybe a dumb idea, but a custom matcher would be nice: `expect(tester.firstName).toHaveValue(person.firstName)`

File: frontend/src/app/person-edit/person-edit.component.spec.ts [hunk 3]
```diff
@@ -228,120 +341,106 @@ describe('PersonEditComponent', () => {
...
       // erase something in the field, which should make its model null
-      city.value = '42000 SAINT-ETIENN';
-      city.dispatchEvent(new Event('input'));
-      fixture.detectChanges();
-      expect(fixture.componentInstance.personForm.value.city).toBeFalsy();
-      expect(city.classList).toContain('is-warning');
+      tester.city.fillWith('42000 SAINT-ETIENN');
+
+      expect(tester.componentInstance.personForm.value.city).toBeFalsy();
+      expect(tester.city.classes).toContain('is-warning');
```
Comment (cexbrayat) [inline frontend/src/app/person-edit/person-edit.component.spec.ts]: Same `tohaveClass` would be cool.

File: frontend/src/app/person-edit/person-edit.component.spec.ts [hunk 4]
```diff
@@ -420,173 +503,94 @@ describe('PersonEditComponent', () => {
...
-      fixture.detectChanges();
-      tick(500);
+      tester.fillAndTick(tester.nationality, 'Bel');
       // select first result
-      const nationalityResult = nativeElement.querySelector('ngb-typeahead-window button');
-      expect(nationalityResult.textContent).toContain('Belgique');
-      nationalityResult.dispatchEvent(new Event('click'));
-      fixture.detectChanges();
-      tick();
+      expect(tester.firstTypeaheadOption.textContent).toContain('Belgique');
```
Comment (cexbrayat) [inline frontend/src/app/person-edit/person-edit.component.spec.ts]: Same: `toContainText` would be nice

File: frontend/src/app/tasks/tasks.component.spec.ts [hunk 5]
```diff
@@ -32,6 +33,44 @@ class TestComponent {
...
+  constructor(fixture: ComponentFixture<TestComponent>) {
+    super(fixture);
+  }
+
+  get text() {
+    return this.element('.task-item').textContent;
+  }
+
+  get addSpentTimeLink() {
+    return this.element('.add-spent-time-link') as TestHtmlElement<HTMLElement>;
```
Comment (cexbrayat) [inline frontend/src/app/tasks/tasks.component.spec.ts]: I don't know if it's better, but I would have tried `this.element<HTMLElement>()`

Comment (jnizet) [reply frontend/src/app/tasks/tasks.component.spec.ts]: I tried a whole lot of things, including that. The less cumbersome way I found was the one I finally picked. I also think it's easier to use for those who aren't too versed into generics.

File: frontend/package.json [hunk 6]
```diff
@@ -54,7 +54,8 @@
     "protractor": "5.3.2",
     "ts-node": "6.0.3",
     "tslint": "5.10.0",
-    "typescript": "2.7.2"
+    "typescript": "2.7.2",
+    "ngx-speculoos": "0.2.0"
```
Comment (cexbrayat) [inline frontend/package.json]: the deps should be in alphabetical order

**Comment by jnizet** [comment]
merged on master

**Pull request closed without merging**